diuf/sudoku/solver/rules/TurbotFishHint.java

```diff
@@ -0,0 +1,211 @@
+        result.put(startCell, fishDigitSet); // orange
+        result.put(bridgeCell1, fishDigitSet);
+        result.put(bridgeCell2, fishDigitSet); // orange
+        result.put(endCell, fishDigitSet);
+        return result;
+    }
+
+    @Override
+    public Map<Cell, BitSet> getRedPotentials(Grid grid, int viewNum) {
+        Map<Cell, BitSet> result = new HashMap<>(super.getRemovablePotentials());
+        BitSet fishDigitSet = SingletonBitSet.create(value);
+        result.put(startCell, fishDigitSet);
+        result.put(bridgeCell2, fishDigitSet);
+        return result;
+    }
+
+    @Override
+    public Collection<Link> getLinks(Grid grid, int viewNum) {
+        Collection<Link> result = new ArrayList<>();
+        result.add(new Link(startCell, value, bridgeCell1, value));
+        result.add(new Link(bridgeCell1, value, bridgeCell2, value));
+        result.add(new Link(bridgeCell2, value, endCell, value));
+        return result;
+    }
+
+    @Override
+    public Grid.Region[] getRegions() {
+        //return new Grid.Region[] { shareRegion };
+		        return null;
+    }
+
+    @Override
+    public String toString() {
+        return getName() +
+                ": " +
+                Cell.toFullString(startCell, bridgeCell1, bridgeCell2, endCell) +
+                " on value " +
+                value;
+    }
+
+    @Override
+    public String toHtml(Grid grid) {
+        String result = HtmlLoader.loadHtml(this, "TurbotFishHint.html");
+        String name = getName();
+        String base = this.baseSet.toFullString();
+        String cover = this.coverSet.toFullString();
+		String shared = this.shareRegion.toFullString();
+        String value = Integer.toString(this.value);
+        String cell1 = startCell.toString();
+        String cell2 = bridgeCell1.toString();
+        String cell3 = bridgeCell2.toString();
+        String cell4 = endCell.toString();
+        result = HtmlLoader.format(result, name, value, cell1, cell2, cell3, cell4, base, cover, shared);
+        return result;
+    }
+
+    @Override
+    public String getName() {
+        Class<? extends Grid.Region> region1 = baseSet.getClass();
+        Class<? extends Grid.Region> region2 = coverSet.getClass();
+        if (region1 == Grid.Row.class) {
+            if (region2 == Grid.Row.class)
+                return "Skyscraper";
+            else
+				if (region2 == Grid.Column.class)
+					return "Two-string Kite";
+				else
+					return "Turbot Fish";
+        }
+		else {
+			if (region1 == Grid.Column.class)
+				if (region2 == Grid.Row.class)
+					return "Two-string Kite";
+				else
+					if (region2 == Grid.Column.class)
+						return "Skyscraper";
+					else 
+                return "Turbot Fish";
+			else
+				return "Turbot Fish";
+        }
+    }	
+	
+    @Override
+    public String getShortName() {
+        Class<? extends Grid.Region> region1 = baseSet.getClass();
+        Class<? extends Grid.Region> region2 = coverSet.getClass();
```

Comment:
Can this (as well as all underlying logic) be replaced by regionTypeIndex?
I removed all similar insanities assuming that additional region constraints are out of scope and sticking to ordinary row/col/block constraints. If you wish to extend the solver to other sudoku variants, it is better to start from older source code.

Reply:
This was an oversight as I expanded on what was there before …
I can see what had been done on other Hint files and amend

Reply:
Sorted

Reply:
Then see if you can turn back `public` to `private` in grid.java (as noted above)
```
    public static final Block[] blocks;
    public static final Row[] rows;
    public static final Column[] columns;
```

Reply:
done

Reply:
Excellent! Thanks!